### DISCUSSION TITLE
Jobs optimization - workflow dependencies

### DISCUSSION THREAD

**Discussion by Behoston** [community/community#26209] [Actions]
<p>I trying to automate releasing python package on pypi. I have two different workflows:</p>
<ol>
<li>Test</li>
<li>Release</li>
</ol>
<p>Tests runs on each push, so it will fire on create release too.</p>
<p>Release run only when release published/created/idk still playing with this action.</p>
<p><strong>I want to test whole code just before release and run tests only once.</strong></p>
<p>Test workflow runs after creating release but I don’t know how to run release after tests and only when not failed. Other reasonable option will be two jobs in one workflow but with different triggers. I can’t find both in docs.</p>
<p>Can somebody tell me how to implement most DRY solution? Partial solution will be import (yaml) test job from test workflow to release workflow and use <code>needs</code> keyword.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/behoston">@behoston</a> ,</p>
<p>You can specify the ‘<a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestbranchestags" rel="nofollow noopener">branches/branches-ignore</a>’ for the ‘push’ event, then create release will not trigger test workflow(push event).  code sample as below:</p>
<pre><code>on:
  push:
    branches:
      - master
</code></pre>
<p>Now the two workflows are seperated.</p>
<p>If you’d like execute test workflow before release workflow, you can add an extra step at the end of test workflow, use <a href="https://developer.github.com/v3/repos/releases/#create-a-release" rel="nofollow noopener">rest api </a>to create release, and if test passed, release workflow will be triggered automatically.</p>
<p>Code sample as below, test workflow yaml:</p>
<pre><code>name: pushevent
on:
  push:
    branches:
      - master
jobs:
  pushevent:
    runs-on: [ubuntu-latest]
    steps:
      - name: test steps
        run: echo test steps...
      - name: create release
        run: |
          curl -v -X POST -u ${{ secrets.PAT }} -H "Accept: application/vnd.github.antiope-preview+json" --data '{"tag_name": "v2.4", "target_commitish": "master", "name": "v2.4"}' https://api.github.com/repos/weide-zhou/ticket17/releases
</code></pre>
<p>The release workflow which will be triggered automatically:</p>
<pre><code>name: releasecheck
on:
  release:
    types:
      - created
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b5c7ce308b70309d55d35bf8972da5a5163d6016.png" data-download-href="/uploads/short-url/pW6vcI91OuxvMx0jxMsHwp46g9U.png?dl=1" title="trigger.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181096063-32d82494-f0ad-415d-90ef-4d679114db47.png" alt="trigger.png" data-base62-sha1="pW6vcI91OuxvMx0jxMsHwp46g9U" width="690" height="103" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b5c7ce308b70309d55d35bf8972da5a5163d6016_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">trigger.png</span><span class="informations">808×121 7.62 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p> Currently it’s not supported to limit the concurrent workflow run numbers, eg: push code 3 times but ran just 1 workflow, it’s not supported.</p>
<p>Thanks.</p>

**Comment by Behoston** [comment]
<p>Finally, I decided to not release automatically, due to release tool that is not ready for CI/CD or does not match my case. But I think disabling test job for version tags will be enough if somebody has similar problem.</p>
